src/WalkingModule/src/Module.cpp

```diff
@@ -598,13 +751,423 @@ bool WalkingModule::updateModule()
+        double leftArmRollError=0;
+        double rightArmRollError=0;
+
+        if (m_useStepAdaptation)
+        {
+            m_isPitchActive=0;
+            m_isRollActive=0;
+
+            for (int var = 0; var < m_robotControlHelper->getActuatedDoFs(); var++)
+            {
+                if(var==3 || var==6)
+                {
+                    leftArmPitchError=leftArmPitchError+ abs(m_qDesired(var)-m_robotControlHelper->getJointPosition()(var));
+                }
+                if(var==4 || var==5)
+                {
+                    leftArmRollError=leftArmRollError+ abs(m_qDesired(var)-m_robotControlHelper->getJointPosition()(var));
+                }
+                if(var==7 || var==10)
+                {
+                    rightArmPitchError=rightArmPitchError+ abs(m_qDesired(var)-m_robotControlHelper->getJointPosition()(var));
+                }
+                if(var==8 || var==9)
+                {
+                    rightArmRollError=rightArmRollError+ abs(m_qDesired(var)-m_robotControlHelper->getJointPosition()(var));
+                }
+            }
+
+            leftArmPitchError=leftArmPitchError+0.1;
+            leftArmRollError=leftArmRollError+0.1;
+            rightArmRollError=rightArmRollError+0.1;
+            rightArmPitchError=rightArmPitchError+0.1;
+
+            if (m_useStepAdaptation)
+            {
+                if (leftArmPitchError>m_stepAdapter->getRollPitchErrorThreshold()(1) )
+                {
+                    TempPitch=leftArmPitchError;
+                    m_isPitchActive=1;
+                }
+                else if( rightArmPitchError>m_stepAdapter->getRollPitchErrorThreshold()(1))
+                {
+                    TempPitch=rightArmPitchError;
+                    m_isPitchActive=1;
+                }
+                else
+                {
+                    TempPitch=0;
+                }
+
+                if (leftArmRollError>m_stepAdapter->getRollPitchErrorThreshold()(0) )
+                {
+                    if (m_leftInContact.front())
+                    {
+                        TempRoll=1*leftArmRollError;
+                    }
+                    if (m_rightInContact.front())
+                    {
+                        TempRoll=-1*leftArmRollError;
+                    }
+                    m_isRollActive=1;
+                }
+                else if( rightArmRollError>m_stepAdapter->getRollPitchErrorThreshold()(0) )
+                {
+                    if (m_rightInContact.front())
+                    {
+                        TempRoll=-1*rightArmRollError;
+                    }
+                    if (m_leftInContact.front())
+                    {
+                        TempRoll=+1*rightArmRollError;
+                    }
+                    m_isRollActive=1;
+                }
+                else
+                {
+                    TempRoll=0;
+                }
+            }
+
```

Comment:
I understand the need to have a working solution, but is there any plan to change this? For the time being, you can create a function somewhere, maybe in the ``StepAdaptation`` module which determines the trigger. This function may tell you if it is the case to step or not.

In addition, with the different ``if`` on ``var``, you are assuming a specific joint order. If you need to add an error manually, what's the point of having a specific threshold? You can simply reduce the threshold.

Reply:
> In addition, with the different if on var, you are assuming a specific joint order. If you need to add an error manually, what's the point of having a specific threshold? You can simply reduce the threshold. 

Honestly, I didn't get your point!
What do you mean about specific threshold? There is some threshold inside the configuration file that we can change it!

Reply:
So do you suggest to put this part inside the step adjustment?

Reply:
> What do you mean about specific threshold? There is some threshold inside the configuration file that we can change it!

In the code you have something like:

```
error = 0;
error = error + getSomeError();
error = error + 0.1;

if (error > threshold) 
    doSomething;
```
I think this is not really clean for a couple of reasons:
- What is the point of doing ``error = error + getSomeError();`` as in https://github.com/MiladShafiee/walking-controllers/blob/e142841733301ad7c98b4577cd4b21fb45067e94/src/WalkingModule/src/Module.cpp#L816 if you know that ``error`` is initialized to zero.
- What is the point of doing ``error = error + 0.1`` as in https://github.com/MiladShafiee/walking-controllers/blob/e142841733301ad7c98b4577cd4b21fb45067e94/src/WalkingModule/src/Module.cpp#L832 if you then compare it to a custom threshold?

I would suggest doing the following instead
```
error = 0;
error = getSomeError();

if (error > threshold2) 
    doSomething;
```
where ``threshold2`` is ``threshold - 0.1``.

> So do you suggest to put this part inside the step adjustment?

Yes, I would hide this part away from ``Module.cpp``. It can become a method called like ``triggerStepAdjustment`` (or the name you prefer) which returns ``true`` if the robot has to perform a step, ``false`` otherwise.

Reply:
> What is the point of doing error = error + getSomeError();

Thanks. Here there was the `error` that is not useful because of the evolution of the code during the developing time.

> What is the point of doing `error = error + 0.1` as in
https://github.com/MiladShafiee/walking-controllers/blob/e142841733301ad7c98b4577cd4b21fb45067e94/src/WalkingModule/src/Module.cpp#L832

At the end the `leftArmPitchError` will be used for both push detection and also instructing a rotation matrix for faking the error of the orientation inside the simple  pendulum estimator. We add for example here the `0.1`  as an offset to completely activating the step adjustment and generating the considerable error for step adjustment. So it's not only the matter of push detection here.

Reply:
You may move also the use of the pendulum estimator in the step adjustment, from which you get directly the corrected CoM position. As you may have noticed, the file Module.cpp is pretty large. Things there get overlooked pretty easily. In principle, ``Module.cpp`` should not contain any logic, but it should simply connect all the blocks. The reason behind this is twofold:
- if you need to modify some behavior, you should modify only the corresponding module, without looking for a line in the mess of the ``Module.cpp``.
- The code in ``Module.cpp`` is shorter.

Long story short, move as much "logic" as you can out of the ``Module.cpp``.

Reply:
I tried to move the mentioned parts inside the step adjustment in the last commit.